### PR TITLE
Cuda: Fix nvcc warnings

### LINKED ISSUE #6991: Several warnings in cuda builds
**Describe the bug**

Several warnings are present in cuda builds, below are for:

### cuda/11.8 + gcc/11.3.0 on Hopper arch (Blake H100 queue):

**TestArray.cpp**
```
01:16:44 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestArray.cpp(54): warning #20013-D: calling a constexpr __host__ function("size") from a __host__ __device__ function("is_equal") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
01:16:44           detected during instantiation of "__nv_bool <unnamed>::is_equal(const L &, const R &) [with L=Kokkos::Array<int, 5UL>, R=Kokkos::Array<int, 5UL>]" 
01:16:44 (70): here
01:16:44 
01:16:44 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestArray.cpp(56): warning #20013-D: calling a constexpr __host__ function("size") from a __host__ __device__ function("is_equal") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
01:16:44           detected during instantiation of "__nv_bool <unnamed>::is_equal(const L &, const R &) [with L=Kokkos::Array<int, 5UL>, R=Kokkos::Array<int, 5UL>]" 
01:16:44 (70): here
01:16:44 
01:16:44 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/src/Kokkos_Array.hpp(143): warning #20013-D: calling a constexpr __host__ function("kokkos_swap") from a __host__ __device__ function("kokkos_swap") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
01:16:44           detected during instantiation of "std::enable_if_t<Kokkos::Impl::is_swappable<U>::value, void> Kokkos::kokkos_swap(Kokkos::Array<<unnamed>::MyInt, 2UL> &, Kokkos::Array<<unnamed>::MyInt, 2UL> &) [with U=<unnamed>::MyInt]" 
```

**TestLegionInitialization.cpp**
```
01:16:50 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestLegionInitialization.cpp(30): warning #177-D: member "<unnamed>::initialization_legion_initialization_Test::test_info_" was declared but never referenced
```

**TestCuda_InterOp_StreamsMultiGPU.cpp**
```
01:17:03 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp(60): warning #177-D: member "<unnamed>::cuda_multi_gpu_managed_views_Test::test_info_" was declared but never referenced
```

**TestCuda_InterOp_StreamsMultiGPU.cpp**
```01:17:03 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp(75): warning #177-D: member "<unnamed>::cuda_multi_gpu_unmanaged_views_Test::test_info_" was declared but never referenced
01:17:03 
01:17:03 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp(99): warning #177-D: member "<unnamed>::cuda_multi_gpu_scratch_space_Test::test_info_" was declared but never referenced
```

**TestParseCmdLineArgsAndEnvVars.cpp**
```
01:17:04 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestParseCmdLineArgsAndEnvVars.cpp(143): warning #177-D: member "<unnamed>::defaultdevicetype_cmd_line_args_num_threads_Test::test_info_" was declared but never referenced
01:17:04 
01:17:04 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestParseCmdLineArgsAndEnvVars.cpp(156): warning #177-D: member "<unnamed>::defaultdevicetype_cmd_line_args_device_id_Test::test_info_" was declared but never referenced
...
```

**TestScopedRegion.cpp**
```
01:18:02 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/tools/TestScopedRegion.cpp(34): warning #177-D: member "<unnamed>::defaultdevicetype_scoped_profile_region_Test::test_info_" was declared but never referenced
```

**Kokkos_Complex.hpp**
```
01:19:50 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/src/Kokkos_Complex.hpp(110): warning #20208-D: 'long double' is treated as 'double' in device code
01:19:50 
01:19:50 Warning #20208-D: 'long double' is treated as 'double' in device code
```

**TestNumericTraits.hpp*
```
01:23:30 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestNumericTraits.hpp(46): warning #20208-D: 'long double' is treated as 'double' in device code
01:23:30 
01:23:30 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestNumericTraits.hpp(104): warning #20208-D: 'long double' is treated as 'double' in device code
01:23:30 
01:23:30 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_cuda_1180_gcc_1130_HOPPER90/kokkos/core/unit_test/TestNumericTraits.hpp(105): warning #20208-D: 'long double' is treated as 'double' in device code
...
# lots more
```

**Please include the following for a minimal reproducer**

**Reproducer** (blake H100 queue):
```bash
source /projects/x86-64-icelake-rocky8/spack-config/blake-setup-user-module-env.sh
module purge
module load cmake gcc/11.3.0 cuda/11.8.0 git

$KOKKOS_PATH/generate_makefile.bash --with-cuda --arch=HOPPER90 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper

make -j16
```

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #6991. Currently, we only turn warnings into errors. We should incorporate all the fixes for the warnings here as well, though.

### CODE REVIEW

File: .jenkins [hunk 1]
```diff
@@ -385,7 +385,7 @@ pipeline {
                               cmake \
                                 -DCMAKE_BUILD_TYPE=Release \
                                 -DCMAKE_CXX_COMPILER=g++-8 \
-                                -DCMAKE_CXX_FLAGS=-Werror \
+                                -DCMAKE_CXX_FLAGS=-Werror -Werror all-warnings \
```
Comment (masterleinad) [inline .jenkins:388]:
```suggestion
                                -DCMAKE_CXX_FLAGS="-Werror -Werror all-warnings" \
```

File: .jenkins [hunk 2]
```diff
@@ -448,7 +448,7 @@ pipeline {
                                 -DCMAKE_BUILD_TYPE=Debug \
                                 -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                                 -DCMAKE_CXX_COMPILER=$WORKSPACE/bin/nvcc_wrapper \
-                                -DCMAKE_CXX_FLAGS=-Werror \
+                                -DCMAKE_CXX_FLAGS=-Werror -Werror all-warnings \
```
Comment (masterleinad) [inline .jenkins:451]:
```suggestion
                                -DCMAKE_CXX_FLAGS="-Werror -Werror all-warnings" \
```

File: .jenkins [hunk 3]
```diff
@@ -345,7 +345,7 @@ pipeline {
                         sh '''rm -rf build && mkdir -p build && cd build && \
                               ../gnu_generate_makefile.bash \
                                 --with-options=compiler_warnings \
-                                --cxxflags="-Werror" \
+                                --cxxflags="-Werror -Werror all-warnings " \
```
[the remark this answers is not in the document]
Comment (dalg24) [reply .jenkins:348]: (apply later when pushing fixes)

File: .jenkins [hunk 4]
```diff
@@ -345,7 +345,7 @@ pipeline {
                         sh '''rm -rf build && mkdir -p build && cd build && \
                               ../gnu_generate_makefile.bash \
                                 --with-options=compiler_warnings \
-                                --cxxflags="-Werror -Werror all-warnings" \
+                                --cxxflags="-Werror -Werror all-warnings -Xcudafe --diag_suppress=20208" \
```
Comment (masterleinad) [inline .jenkins:348]: I needed to move pragma push into the `Complex.hpp` to avoid seeing the warning. At this point, it seems to be better to just disable the warning for this build in the compiler flags.

File: core/src/Cuda/Kokkos_Cuda.hpp [hunk 5]
```diff
@@ -186,7 +186,7 @@ class Cuda {
   ///
   /// This matches the __CUDA_ARCH__ specification.
   KOKKOS_DEPRECATED static size_type device_arch() {
-    const cudaDeviceProp& cudaProp = Cuda().cuda_device_prop();
+    const cudaDeviceProp cudaProp = Cuda().cuda_device_prop();
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda.hpp:189]: What warning was raised here?

File: tpls/gtest/gtest/gtest.h [hunk 6]
```diff
@@ -4910,7 +4910,7 @@ class NeverThrown {
   class GTEST_TEST_CLASS_NAME_(test_suite_name, test_name)                    \
       : public parent_class {                                                 \
    public:                                                                    \
-    GTEST_TEST_CLASS_NAME_(test_suite_name, test_name)() = default;           \
+    GTEST_TEST_CLASS_NAME_(test_suite_name, test_name)() { (void)test_info_; }\
```
Comment (dalg24) [inline tpls/gtest/gtest/gtest.h:4913]: Could you please reference issues or rejected prs on googletest that show that there is no interest on resolving this upstream?

File: core/src/Cuda/Kokkos_Cuda.hpp (hunk 5 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda.hpp:189]: dangling reference

File: tpls/gtest/gtest/gtest.h (hunk 6 above)
Comment (masterleinad) [reply tpls/gtest/gtest/gtest.h:4913]: https://github.com/google/googletest/issues/4104#issuecomment-1376301742 mentions that `nvcc` is an unsupported compiler (in a slightly different context).

**Comment by masterleinad** [comment]
Retest this please.

**Review by crtrott — APPROVED** [review]

**Comment by Rombur** [comment]
Retest this please

**Comment by masterleinad** [comment]
The only errors in Cuda builds are
```
var/jenkins/workspace/Kokkos_PR-7021/core/unit_test/TestTeamMDRangePolicyCTAD.cpp(76): error #177-D: function "<unnamed>::TestTeamThreadMDRangeCTAD::test_ctad_inside_parallel_for" was declared but never referenced
/var/jenkins/workspace/Kokkos_PR-7021/core/unit_test/TestTeamMDRangePolicyCTAD.cpp(136): error #177-D: function "<unnamed>::TestTeamVectorMDRangeCTAD::test_ctad_inside_parallel_for" was declared but never referenced
/var/jenkins/workspace/Kokkos_PR-7021/core/unit_test/TestTeamMDRangePolicyCTAD.cpp(196): error #177-D: function "<unnamed>::TestThreadVectorMDRangeCTAD::test_ctad_inside_parallel_for" was declared but never referenced
```
fixed by #7049.

**Comment by masterleinad** [comment]
Rebased after #7049 was merged.

File: core/unit_test/TestNumericTraits.hpp [hunk 7]
```diff
@@ -21,6 +21,19 @@
...
+#ifdef __NVCC_DIAG_PRAGMA_SUPPORT__
+#pragma nv_diagnostic push
+#pragma nv_diag_suppress 20208
+#else
+#ifdef __CUDA_ARCH__
+#pragma diagnostic push
+#pragma diag_suppress 20208
+#endif
+#endif
+#endif
```
Comment (dalg24) [inline core/unit_test/TestNumericTraits.hpp:35]: I feel that is not the right approach for that test.
We already have all sorts of guards for `long double` in these, I feel like we should consider reworking them like we did for the math function (I don;t remember the details but I see that it is not part of this PR meaning we somehow resolved it without suppressing the diagnostic altogether)

File: tpls/gtest/gtest/gtest.h (hunk 6 above)
Comment (masterleinad) [reply tpls/gtest/gtest/gtest.h:4913]: > Could it be a problem if we later want to update gtest ?

We will notice if this is still a problem or not.

> What do you think of pre-installing all Kokkos dependencies in the CI so that we use the flags we want solely on Kokkos ? We could then compile gtest with the host compiler ?

This is not about installing `gtest` but rather using it. We compiler warnings when we define unit tests.

File: core/unit_test/TestNumericTraits.hpp (hunk 7 above)
Comment (masterleinad) [reply core/unit_test/TestNumericTraits.hpp:35]: We could just disable all tests for `long double` when we compile with `Cuda` support (which is what we are doing for other configurations) but these seem to work fine after suppressing the warning.

**Comment by masterleinad** [comment]
Only `'HIP-ROCm-5.6-C++20'` is failing.

File: tpls/gtest/gtest/gtest.h (hunk 6 above)
Comment (tpadioleau) [reply tpls/gtest/gtest/gtest.h:4913]: > This is not about installing `gtest` but rather using it. We compiler warnings when we define unit tests.

I had in mind that using an installed gtest would treat gtest include header path as system path thus removing the warnings ?

Comment (masterleinad) [reply tpls/gtest/gtest/gtest.h:4913]: We could try that (in a separate pull request); it's certainly a bigger change than changing this one line.

**Review by aprokop — APPROVED** [review]
Looks fine to me.

File: core/unit_test/TestNumericTraits.hpp (hunk 7 above)
Comment (ajpowelsnl) [reply core/unit_test/TestNumericTraits.hpp:35]: Are the `long double` tests on CUDA already effectively switched off, in that if they are run as `double` in CUDA device code?

**Review by ajpowelsnl — COMMENTED** [review]
These fixes look fine, however, questions remain about more (potentially) complicated fixes for warnings associated with methods in `TestStdAlgorithmsCommon.hpp` that appear in CUDA-11.2.x compilations -- do we want to fix warnings that will likely disappear with CUDA updates?

**Comment by masterleinad** [comment]
> These fixes look fine, however, questions remain about more (potentially) complicated fixes for warnings associated with methods in `TestStdAlgorithmsCommon.hpp` that appear in CUDA-11.2.x compilations -- do we want to fix warnings that will likely disappear with CUDA updates?

The changes here are sufficient for the configurations in the CI. We can discuss other configurations in the original issue or a follow-up. nvcc 11.0 in particular will raise quite a bit more warnings and it's not clear to me if that's worth putting a lot of effort in (which might make code less readable).

File: core/unit_test/TestNumericTraits.hpp (hunk 7 above)
Comment (masterleinad) [reply core/unit_test/TestNumericTraits.hpp:35]: Yes, everything that uses `TestNumericTraits` at least, see https://github.com/kokkos/kokkos/blob/63f05204d0eaeec83e25f87eeb8025a188683f92/core/unit_test/TestNumericTraits.hpp#L182-L204.

**Comment by ajpowelsnl** [comment]
> > These fixes look fine, however, questions remain about more (potentially) complicated fixes for warnings associated with methods in `TestStdAlgorithmsCommon.hpp` that appear in CUDA-11.2.x compilations -- do we want to fix warnings that will likely disappear with CUDA updates?
> 
> The changes here are sufficient for the configurations in the CI. We can discuss other configurations in the original issue or a follow-up. nvcc 11.0 in particular will raise quite a bit more warnings and it's not clear to me if that's worth putting a lot of effort in (which might make code less readable).

OK, sounds good.  I'll see if there are any more "low hanging fruit" fixes to do in the issue.

**Comment by ndellingwood** [comment]
Nice fixes :) 
I added a changelog entry for 4.4 under General Enhancements, but please move elsewhere if deemed more appropriate (Bug Fixes?)
